Fixes #36251 - Add applied_filters field in CV version table and store filters during publish

sjha4 opened the pull request:
#### What are the changes introduced in this pull request?

#### Considerations taken when implementing this change?

#### What are the testing steps for this pull request?

qcjames53 commented:
Everything looked great in my testing; I tried every filter type with several different combinations each. The filter, rules, affected repos section for each filter (along with the dependency solving flag) always updated appropriately on publish. Source code looked good; accepting.

Thanks for all the help with this one, Samir!

parthaa commented on app/lib/actions/katello/content_view/publish.rb:
```diff
@@ -110,6 +110,7 @@ def rescue_strategy_for_self
         def finalize
           version = ::Katello::ContentViewVersion.find(input[:content_view_version_id])
           version.update_content_counts!
+          version.add_applied_filters
```
```suggestion
          version.add_applied_filters!
```

parthaa replied:
We use ! to indicate methods that will change the object

parthaa commented:
Have you thought of adding a "serialize" (for example https://github.com/Katello/katello/blob/master/app/models/katello/content_view_version.rb#L71) or does that not work for jsonb?

parthaa commented:
Should we also update the content view version show json rabl to include data along "cv applied" or otherwise.

sjha4 commented:
> We use ! to indicate methods that will change the object

Updated 👍🏼 

> Have you thought of adding a "serialize" (for example https://github.com/Katello/katello/blob/master/app/models/katello/content_view_version.rb#L71) or does that not work for jsonb?

Updated to use Rails default serialization for jsonb fields.

> Should we also update the content view version show json rabl to include data along "filters applied" or otherwise.

There's a separate task for that. We will be adding the rabl changes there.

parthaa commented on app/models/katello/content_view_version.rb:
```diff
@@ -428,6 +428,17 @@ def after_promote_hooks
 
+    def add_applied_filters!
+      applied_filters_and_rules = content_view.filters.map do |f|
+        { filter: f, rules: f.rules, affected_repos: f.repositories }
```
1. What metadata of the repo gets stored ?
2. What happens if your filter did not have any repos (like we assume its for all repos? )

sjha4 replied:
> What metadata of the repo gets stored ?

```
[2] pry(main)> Katello::ContentViewVersion.find(55).applied_filters
=> {"applied_filters"=>
  [{"rules"=>
     [{"id"=>3, "name"=>"bear", "version"=>nil, "created_at"=>"2023-03-31T14:37:26.548-04:00", "updated_at"=>"2023-03-31T14:37:26.548-04:00", "max_version"=>nil, "min_version"=>nil, "architecture"=>"", "content_view_filter_id"=>9}],
    "filter"=>
     {"id"=>9,
      "name"=>"prod",
      "repos"=>["needed_errata", "modular"],
      "content"=>"rpm",
      "inclusion"=>true,
      "created_at"=>"2023-03-31T14:36:24.750-04:00",
      "updated_at"=>"2023-03-31T14:37:08.196-04:00",
      "description"=>"",
      "organization"=>"Default_Organization",
      "content_view_id"=>7,
      "original_packages"=>true,
      "content_view_label"=>"cv11",
      "original_module_streams"=>false},
    "affected_repos"=>
     [{"id"=>1,
       "pulp_id"=>"5d4ee771-1a85-405b-9402-f91be9a4e121",
       "root_id"=>1,
       "remote_href"=>"/pulp/api/v3/remotes/rpm/rpm/3ed45878-58b2-4a3b-bfcd-bfe47a7329dd/",
       "last_indexed"=>"2023-03-22T11:55:12.390-04:00",
       "version_href"=>"/pulp/api/v3/repositories/rpm/rpm/55af24b6-76f9-4dcb-92a7-d500a69c6304/versions/1/",
       "relative_path"=>"Default_Organization/Library/custom/prod/needed_errata",
       "environment_id"=>1,
       "publication_href"=>"/pulp/api/v3/publications/rpm/rpm/f93bb4a7-7658-4775-b069-efa6191ec1d6/",
       "distribution_arch"=>nil,
       "distribution_family"=>nil,
       "library_instance_id"=>nil,
       "saved_checksum_type"=>nil,
       "distribution_variant"=>nil,
       "distribution_version"=>nil,
       "distribution_bootable"=>nil,
       "last_contents_changed"=>"2023-03-22T11:55:09.813-04:00",
       "content_view_version_id"=>1,
       "last_applicability_regen"=>"2023-03-22T11:55:12.489-04:00",
       "container_repository_name"=>nil},
      {"id"=>5,
       "pulp_id"=>"135fd767-2e9f-431c-8994-9460df7653d0",
       "root_id"=>2,
       "remote_href"=>"/pulp/api/v3/remotes/rpm/rpm/73a03155-7f36-470a-ae66-b21fc1afa878/",
       "last_indexed"=>"2023-04-03T10:01:28.457-04:00",
       "version_href"=>"/pulp/api/v3/repositories/rpm/rpm/fd833b5e-d050-4dba-aff2-b8651b10b2bc/versions/4/",
       "relative_path"=>"Default_Organization/Library/custom/prod/modular",
       "environment_id"=>1,
       "publication_href"=>"/pulp/api/v3/publications/rpm/rpm/a7e2e165-6168-4bc3-a149-ef1d8da5b81a/",
       "distribution_arch"=>"x86_64",
       "distribution_family"=>"Test Family",
       "library_instance_id"=>nil,
       "saved_checksum_type"=>nil,
       "distribution_variant"=>"TestVariant",
       "distribution_version"=>"16",
       "distribution_bootable"=>false,
       "last_contents_changed"=>"2023-04-03T10:01:25.182-04:00",
       "content_view_version_id"=>1,
       "last_applicability_regen"=>"2023-04-03T10:01:28.576-04:00",
       "container_repository_name"=>nil}]}],
 "dependency_solving"=>false}

```

>What happens if your filter did not have any repos (like we assume its for all repos? )

It stores affected_repos as an empty array.
```
[1] pry(main)> Katello::ContentViewVersion.find(54).applied_filters
=> {"applied_filters"=>
  [{"rules"=>
     [{"id"=>3, "name"=>"bear", "version"=>nil, "created_at"=>"2023-03-31T14:37:26.548-04:00", "updated_at"=>"2023-03-31T14:37:26.548-04:00", "max_version"=>nil, "min_version"=>nil, "architecture"=>"", "content_view_filter_id"=>9}],
    "filter"=>
     {"id"=>9,
      "name"=>"prod",
      "repos"=>[],
      "content"=>"rpm",
      "inclusion"=>true,
      "created_at"=>"2023-03-31T14:36:24.750-04:00",
      "updated_at"=>"2023-03-31T14:37:08.196-04:00",
      "description"=>"",
      "organization"=>"Default_Organization",
      "content_view_id"=>7,
      "original_packages"=>true,
      "content_view_label"=>"cv11",
      "original_module_streams"=>false},
    "affected_repos"=>[]}],
 "dependency_solving"=>false}

```

parthaa commented on app/models/katello/content_view_version.rb:
```diff
@@ -428,6 +428,24 @@ def after_promote_hooks
+          rules: f.rules,
+          affected_repos: f.applicable_repos.map do |repo|
+            repo.slice(:id, :name, :label, :arch, :major, :minor,
+                       :content_type, :os_versions, :url, :content_id).merge(redhat: repo.redhat?, root: repo.root.id)
```
```suggestion
                       :content_type, :os_versions, :url, :content_id).merge(redhat: repo.redhat?, root: repo.root.id, product: repo.product.slice(:id, :name, :label))
```

sjha4 replied:
Updated to add product id and label. Here's what it looks like now:

```
{"applied_filters"=>
  [{"rules"=>
     [{"id"=>3, "name"=>"bear", "version"=>nil, "created_at"=>"2023-03-31T14:37:26.548-04:00", "updated_at"=>"2023-03-31T14:37:26.548-04:00", "max_version"=>nil, "min_version"=>nil, "architecture"=>"", "content_view_filter_id"=>9}],
    "filter"=>
     {"id"=>9,
      "name"=>"prod",
      "repos"=>["needed_errata", "modular"],
      "content"=>"rpm",
      "inclusion"=>true,
      "created_at"=>"2023-03-31T14:36:24.750-04:00",
      "updated_at"=>"2023-03-31T14:37:08.196-04:00",
      "description"=>"",
      "organization"=>"Default_Organization",
      "content_view_id"=>7,
      "original_packages"=>true,
      "content_view_label"=>"cv11",
      "original_module_streams"=>false},
    "affected_repos"=>
     [{"id"=>1,
       "url"=>"https://jlsherrill.fedorapeople.org/fake-repos/needed-errata/",
       "arch"=>"noarch",
       "name"=>"needed_errata",
       "root"=>1,
       "label"=>"needed_errata",
       "major"=>nil,
       "minor"=>nil,
       "redhat"=>false,
       "product"=>{"id"=>1, "label"=>"prod"},
       "content_id"=>"1679500443174",
       "os_versions"=>[],
       "content_type"=>"yum"},
      {"id"=>5,
       "url"=>"https://partha.fedorapeople.org/test-repos/pteradactyl/",
       "arch"=>"noarch",
       "name"=>"modular",
       "root"=>2,
       "label"=>"modular",
       "major"=>nil,
       "minor"=>nil,
       "redhat"=>false,
       "product"=>{"id"=>1, "label"=>"prod"},
       "content_id"=>"1679504607274",
       "os_versions"=>[],
       "content_type"=>"yum"}]},
   {"rules"=>
     [{"id"=>56,
       "types"=>[],
       "end_date"=>nil,
       "date_type"=>"updated",
       "errata_id"=>"RHEA-2013:0075",
       "created_at"=>"2023-04-05T09:49:07.999-04:00",
       "start_date"=>nil,
       "updated_at"=>"2023-04-05T09:49:07.999-04:00",
       "content_view_filter_id"=>10},
      {"id"=>57,
       "types"=>[],
       "end_date"=>nil,
       "date_type"=>"updated",
       "errata_id"=>"RHEA-2013:0070",
       "created_at"=>"2023-04-05T09:49:08.005-04:00",
       "start_date"=>nil,
       "updated_at"=>"2023-04-05T09:49:08.005-04:00",
       "content_view_filter_id"=>10}],
    "filter"=>
     {"id"=>10,
      "name"=>"needed_errata",
      "repos"=>["zoo", "modular"],
      "content"=>"erratum",
      "inclusion"=>false,
      "created_at"=>"2023-04-05T09:49:02.184-04:00",
      "updated_at"=>"2023-04-05T09:49:02.184-04:00",
      "description"=>"",
      "organization"=>"Default_Organization",
      "content_view_id"=>7,
      "original_packages"=>false,
      "content_view_label"=>"cv11",
      "original_module_streams"=>false},
    "affected_repos"=>
     [{"id"=>6,
       "url"=>"https://partha.fedorapeople.org/test-repos/zoo1/",
       "arch"=>"noarch",
       "name"=>"zoo",
       "root"=>3,
       "label"=>"zoo",
       "major"=>nil,
       "minor"=>nil,
       "redhat"=>false,
       "product"=>{"id"=>1, "label"=>"prod"},
       "content_id"=>"1679504669458",
       "os_versions"=>[],
       "content_type"=>"yum"},
      {"id"=>5,
       "url"=>"https://partha.fedorapeople.org/test-repos/pteradactyl/",
       "arch"=>"noarch",
       "name"=>"modular",
       "root"=>2,
       "label"=>"modular",
       "major"=>nil,
       "minor"=>nil,
       "redhat"=>false,
       "product"=>{"id"=>1, "label"=>"prod"},
       "content_id"=>"1679504607274",
       "os_versions"=>[],
       "content_type"=>"yum"}]}],
 "dependency_solving"=>false}
```

sjha4 commented:
[test katello]

parthaa reviewed (approved):
Works well for me
